### ISSUE TITLE
Cannot Reproduce Results on Retina dataset.

### ISSUE THREAD

**Issue by sadimanna** [rui-yan/SSL-FL#22]
The code given in the repo seems to work well on the Central split of Retina. But when trained on Split 2, the results seem to be eerily strange. The model is predicting only 1s in the FT validation step.

The loss in the training is stuck at *0.61*, even after *1600* epochs of pre-training. I am using a BS of 128 (64 $\times$ 2 GPUs) for pre-training with a base LR of 2e-3, and a mask ratio of 0.6. The optimizer is AdamW. 

I noticed the $\beta_2$ hyper-param in the code is 0.95, whereas in the paper it is written to be 0.999. I changed 0.95 to 0.999 only. The warmup epochs were written to be 5, but in the '.sh', it was written to be 40 for 1600 pre-training epochs. I wonder if that influenced the performance.

For FT, the LR is 3e-3, with a BS of 256 (128 $\times$ 2 GPUs). Here also, the loss gets stuck after a few epochs, and the validation accuracy does not improve, stuck at 50%. This phenomenon is also observed when trying to implement FedBYOL using the hyper-parameters mentioned in the paper on the Retina dataset.

However, on the Central split, the same hyper-parameter config yields **73.67**% accuracy only after 500 epochs of pre-training. I believe that will surely improve if pre-trained for 1600 epochs. But why is there a discrepancy when pre-training on Retina Split-2 with simple FedAvg? Even fine-tuning the ViT on Central split from scratch (without SSL pre-training) conforms with the results given in the paper, but it again failed to reproduce the given result on Split-2.

Kindly, provide some hints regarding what can go wrong when pre-training with Split-2.

**UPDATE**: 
Accuracy: 81.73% on Derm Split-2, Pre-training 500 epochs, BLR 2e-3, BS 128, Warm-Up epochs 10 
FTLR 3e-3, FTBS 256, FT epochs 100

**Comment by rui-yan** [comment]
Hi, 

Below are the reported results on the Retina dataset:

![Screenshot 2024-11-11 at 11 45 27 AM](https://github.com/user-attachments/assets/71d6ae49-fd14-4249-b45c-34c7f67879c4)

Could you confirm if the results for Retina Split-2 after fine-tuning are consistently low across all methods? This includes approaches without pre-training, using a pre-trained model on ImageNet as well as our proposed SSL pre-training methods: Fed-BEiT and Fed-MAE.

Additionally, when you mentioned that 'The training loss is stuck at 0.61, even after 1600 epochs of pre-training,' is this referring to Fed-BEiT or Fed-MAE?

**Comment by sadimanna** [comment]
@rui-yan 

I ran both FedMAE and FedBYOL on Retina dataset, on both the Central and Split-2. While I do not face this issue in Central split. I face this issue only for the Split-2 for both FedMAE and FedBYOL.

The 1600 epochs pre-training was for FedMAE.

It is unlikely, that the same framework will work on Central and fail to work on Split-2 by that margin. I inspected the losses, and the loss for each client decreases by a certain amount from the starting loss value, but the loss of the validation epochs is stuck around 0.693 (~ ln(2) ) which gives 50% accuracy. This is a very strange phenomenon.

I have even tried with an FT LR as low as 5e-5, but the same thing occurs. And it occurs on repeated runs and only the **non-IID split of Retina**, but this does not occur for Derm or COVID-FL non-IID splits.

Hence, the call for hints.

Will update if I get better results.

**Comment by rui-yan** [comment]
For your reference, here is the training loss I obtained while pre-training on the Retina Split-2 dataset using Fed-MAE: https://drive.google.com/file/d/171nmZjPO1lq7Oygu2b1KPfJ0CopDzhNS/view?usp=sharing. The training loss reached 0.61 within the first 100 pre-training epochs.

**Comment by sadimanna** [comment]
@rui-yan Thank you for providing the loss values. In my experiment, the loss reaches 0.61 like yours, but after that it does not decrease further.

I think I will recheck the HPs and run it again. Thank you for your help.
